File: src/main/java/RandomCipher.java
import java.util.HashMap;
import java.util.Map;
import java.util.Random;

class RandomCipher {
  private static String message = "Meet me at secret location at noon on Wednesday.".toLowerCase();
  private static Random random = new Random();

  public static void main(String[] args) {
    System.out.println(hardCipher());
  }

  static String hardCipher() {
    Map<Character, Character> cipher = new HashMap<Character, Character>();
    String uniqueCharacters = "";

    for (char c : message.toCharArray())
      if (Character.isLetterOrDigit(c) && !uniqueCharacters.contains(c + ""))
        uniqueCharacters += c;

    // Map each letter to a random letter in the alphabet (n.b. one-to-one)
    String alphabet = "abcdefghijklmnopqrstuvwxyz";
    for (char c : uniqueCharacters.toCharArray()) {
      char x = alphabet.charAt(random.nextInt(alphabet.length()));
      cipher.put(c, x);
      alphabet = alphabet.replace(x + "", "");
    }

    StringBuilder sb = new StringBuilder();
    for (int i = 0; i < message.length(); i++) {
      char c = message.charAt(i);
      if (cipher.containsKey(c))
        sb.append(cipher.get(c));
      else
        sb.append(c);
    }

    return sb.toString();
  }
}

File: src/main/kotlin/RandomCipher.kt
import java.security.SecureRandom
import java.util.*

/**
 * The random cipher, or mixed alphabet cipher, maps each letter in the alphabet with cipherWord randomly
 * chosen letter of the same alphabet, without replacement (ie. A->M, B->M is not allowed). This
 * cipher is also commonly referred to as cipherWord "monoalphabetic substitution cipher".
 */

fun main(args: Array<String>) {
    val message = "four score and seven years ago our fathers brought forth to this continent a new nation conceived in liberty and dedicated".toLowerCase()

    println(hardCipher(message))
}

fun hardCipher(message: String): String {
    val random = SecureRandom()
    val cipher = HashMap<Char, Char>()
    var uniqueCharacters = ""

    for (c in message.toCharArray())
        if (Character.isLetterOrDigit(c) && !uniqueCharacters.contains(c + ""))
            uniqueCharacters += c

    // Map each letter to cipherWord random letter in the alphabet (n.b. one-to-one)
    var alphabet = "abcdefghijklmnopqrstuvwxyz"
    for (c in uniqueCharacters.toCharArray()) {
        val x = alphabet[random.nextInt(alphabet.length)]
        cipher.put(c, x)
        alphabet = alphabet.replace(x + "", "")
    }

    val sb = StringBuilder()
    for (i in 0..message.length - 1) {
        val c = message[i]
        if (cipher.containsKey(c))
            sb.append(cipher[c])
        else
            sb.append(c)
    }

    return sb.toString()
}